File: C/power.c
#include<stdio.h>
float power(int x, int y){
    if (y == 0){
        return 1;
    }
    else if ( y >0){
        return x*power(x, y-1);
    }
    else{
        return 1.0/x*power(x,y+1);
    }
}
int main() {
    int base, exponent;

    printf("Enter the base: ");
    scanf("%d", &base);

    printf("Enter the exponent: ");
    scanf("%d", &exponent);

    float result = power(base, exponent);

    printf("%d ^ %d = %f\n", base, exponent, result);

    return 0;
}

File: DSA/recursion_dp/power.py
# Recursion
def power(base, exponent):
    assert exponent >=0 and int(exponent) == exponent, "The number should be a whole number"
    if exponent == 0:
        return 1
    elif exponent>0:
        return base * power(base, exponent-1)
    else:
        return 1/base * power(base, -exponent+1)
# Time Complexity : O(n)
# Space Complexity: O(n)


# binary exponentiation algorithm
def power(base, exponent):
    assert exponent >=0 and int(exponent) == exponent, "The number should be a whole number"
    result = 1
    while exponent > 0:
        if exponent % 2 == 1:
            result *= base
        base *= base
        exponent = exponent // 2
    return result
# Time Complexity : O(log(n))
# Space Complexity: O(1)